File: docs/examples/cpp/factory_pattern.cpp
// Factory Pattern usage examples.

// --------------------------------------------------------------------------------------------
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE.txt in the project root for
// license information.
// --------------------------------------------------------------------------------------------

// start-cell-1
#include <qdk/chemistry.hpp>

using namespace qdk::chemistry::algorithms;

// Create default implementation
auto scf_solver = ScfSolverFactory::create();

// Create specific implementation by name
auto localizer = LocalizerFactory::create("pipek-mezey");

// Configure and use the instance
scf_solver->settings().set("basis_set", "def2-tzvp");
auto [E_scf, orbitals] = scf_solver->solve(structure);
// end-cell-1

// start-cell-2
#include <external_program/api.h>  // Your external program's API

#include <qdk/chemistry/algorithms/scf_solver.hpp>

namespace my_namespace {

class ExternalProgramScfSolver : public qdk::chemistry::algorithms::ScfSolver {
 public:
  ExternalProgramScfSolver() = default;
  ~ExternalProgramScfSolver() override = default;

  // Implement the interface method that connects to your external program
  std::tuple<double, qdk::chemistry::data::Orbitals> solve(
      const qdk::chemistry::data::Structure& structure) override {
    // Convert QDK/Chemistry structure to external program format
    auto ext_molecule = convert_to_external_format(structure);

    // Run calculation using external program's API
    auto ext_results =
        external_program::run_scf(ext_molecule, settings().get_all());

    // Convert results back to QDK/Chemistry format
    double energy = ext_results.energy;
    qdk::chemistry::data::Orbitals orbitals =
        convert_from_external_format(ext_results.orbitals);

    return {energy, orbitals};
  }

 private:
  // Helper functions for format conversion
  external_program::Molecule convert_to_external_format(
      const qdk::chemistry::data::Structure& structure);
  qdk::chemistry::data::Orbitals convert_from_external_format(
      const external_program::Orbitals& ext_orbitals);
};

}  // namespace my_namespace
// end-cell-2

File: docs/examples/python/factory_pattern.py
"""Factory pattern usage examples."""

# --------------------------------------------------------------------------------------------
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE.txt in the project root for license information.
# --------------------------------------------------------------------------------------------

import numpy as np
from qdk_chemistry.data import Structure

# Create a simple molecule for testing
coords = np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 1.4]])
structure = Structure(coords, ["H", "H"])

# Create algorithms using the factory
# start-cell-1
from qdk_chemistry.algorithms import create

# Create default implementation
scf_solver = create("scf_solver")

# Create specific implementation by name
localizer = create("orbital_localizer", "qdk_pipek_mezey")

# Configure and use the instance
scf_solver.settings().set("basis_set", "cc-pvdz")
E_scf, wfn = scf_solver.run(structure, charge=0, spin_multiplicity=1)
# end-cell-1

# List available algorithms
# print("Available SCF solvers:", available("scf_solver"))
# print("Available Hamiltonian constructors:", available("hamiltonian_constructor"))
# print("Available orbital localizers:", available("orbital_localizer"))

# # Example of creating with custom settings
# scf_solver = create("scf_solver")
# scf_solver.settings().set("max_iterations", 100)
# scf_solver.settings().set("basis_set", "sto-3g")
# print(f"Set max_iterations to: {scf_solver.settings().get('max_iterations')}")

# # Run the SCF solver with the molecule (requires resources directory)
# # E_scf, wfn = scf_solver.run(molecule, charge=0, spin_multiplicity=1)
# # print(f"SCF Energy: {E_scf}")
